From bd231aefbe41cf86821245da032d7deff84e34b0 Mon Sep 17 00:00:00 2001
From: JayZhang-wiwynn <jay_zhang@wiwynn.com>
Date: Thu, 11 Jul 2024 15:30:32 +0800
Subject: [PATCH 1/4] proprietary: intel: BHS crashdump v0.9 to support without
 systemd/dbus

Description:
 BHS crashdump v0.9 to support without systemd/dbus

Motivation:
 Support some system without dbus
---
 CMakeLists.txt            | 25 +++++++++++++++----------
 PlatInfo.cpp              | 39 +++++++++++++++++++++++++++++++++++++++
 crashdump.cpp             | 25 +++++++++++++++++++++++++
 crashdump.hpp             |  6 ++++++
 utils_common_dbusplus.cpp | 23 +++++++++++++++++++++++
 utils_common_dbusplus.hpp |  4 ++++
 6 files changed, 112 insertions(+), 10 deletions(-)
 create mode 100644 PlatInfo.cpp

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 12ce7c2..813b605 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -15,7 +15,7 @@
 
 
 cmake_minimum_required (VERSION 3.6)
-option (YOCTO_DEPENDENCIES "Use YOCTO dependencies system" OFF)
+option (YOCTO_DEPENDENCIES "Use YOCTO dependencies system" ON)
 option (CRASHDUMP_BUILD_UT "Enable Building Unit tests" OFF)
 
 # optional features
@@ -41,7 +41,7 @@ endif ()
 # Set the searching location for cmake 'include' locations
 set (CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}")
 # Include the sanitizer module
-include (code-coverage)
+#include (code-coverage)
 
 set (CMAKE_CXX_STANDARD 20)
 set (CMAKE_CXX_STANDARD_REQUIRED ON)
@@ -124,7 +124,6 @@ if (NOT ${YOCTO_DEPENDENCIES})
         PATCH_COMMAND
             git apply
             ${CMAKE_BINARY_DIR}/../../openbmc-meta-intel/meta-common/recipes-core/libpeci/libpeci/0002-Add-PECI-Telemetry-command.patch
-            ${CMAKE_BINARY_DIR}/../../openbmc-meta-intel/meta-common/recipes-core/libpeci/libpeci/0001-Enable-64-bit-RdPkgConfig-and-WrPkgConfig-transfers.patch
         LOG_DOWNLOAD ON
     )
     externalproject_get_property (libpeci-project SOURCE_DIR)
@@ -207,6 +206,7 @@ set (
     engine/consolelog.c
     engine/tpmi.c
     engine/modules.c
+    PlatInfo.cpp
 )
 
 set (UTILS_COMMON_FILE utils_common_dbusplus.cpp)
@@ -264,25 +264,28 @@ if (NOT ${CRASHDUMP_BUILD_UT})
         crashdump
         PRIVATE $<$<BOOL:${CRASHDUMP_PRINT_UNFORMATTED}>:
                 -DCRASHDUMP_PRINT_UNFORMATTED> $<$<BOOL:${OEMDATA_SECTION}>:
-                -DOEMDATA_SECTION>
+                -DOEMDATA_SECTION> $<$<BOOL:${NO_SYSTEMD}>: -DNO_SYSTEMD>
     )
 
+    if (NOT ${NO_SYSTEMD})
+        set (DBUS_LIB sdbusplus -lsystemd)
+    endif ()
+
     target_include_directories (crashdump PRIVATE ${CMAKE_SOURCE_DIR})
     set_property (TARGET crashdump PROPERTY C_STANDARD 99)
     target_link_libraries (
         crashdump
         cjson
         peci
-        -lsystemd
         pthread
         -lstdc++fs
-        sdbusplus
         -ldl
         ${SAFEC_LIBRARIES}
         ${NVD_LIB}
         ${PMT_LIB}
         ${PMT_MCTP_LIB}
         ${BOOST_COROUTINE}
+        ${DBUS_LIB}
     )
     # Remote Module Shared Object
     if (${REMOTE_MODULE_EXAMPLE})
@@ -432,8 +435,10 @@ endif ()
 find_package (PkgConfig REQUIRED)
 pkg_get_variable(SYSTEMD_SYSTEM_UNIT_DIR systemd systemdsystemunitdir)
 
-set (SERVICE_FILES
-     ${PROJECT_SOURCE_DIR}/service_files/com.intel.crashdump.service
-)
+if (NOT ${NO_SYSTEMD})
+    set (SERVICE_FILES
+         ${PROJECT_SOURCE_DIR}/service_files/com.intel.crashdump.service
+    )
 
-install (FILES ${SERVICE_FILES} DESTINATION "${SYSTEMD_SYSTEM_UNIT_DIR}")
+    install (FILES ${SERVICE_FILES} DESTINATION "${SYSTEMD_SYSTEM_UNIT_DIR}")
+endif ()
diff --git a/PlatInfo.cpp b/PlatInfo.cpp
new file mode 100644
index 0000000..a191cdf
--- /dev/null
+++ b/PlatInfo.cpp
@@ -0,0 +1,39 @@
+/******************************************************************************
+ *
+ * INTEL CONFIDENTIAL
+ *
+ * Copyright 2019 Intel Corporation.
+ *
+ * This software and the related documents are Intel copyrighted materials,
+ * and your use of them is governed by the express license under which they
+ * were provided to you ("License"). Unless the License provides otherwise,
+ * you may not use, modify, copy, publish, distribute, disclose or transmit
+ * this software or the related documents without Intel's prior written
+ * permission.
+ *
+ * This software and the related documents are provided as is, with no express
+ * or implied warranties, other than those that are expressly stated in the
+ * License.
+ *
+ ******************************************************************************/
+
+extern "C" {
+#include <stdio.h>
+}
+#include <string>
+
+void getBmcVersion(char *ver, size_t) {
+  (void)ver;
+}
+
+void getBiosVersion(char *ver, size_t) {
+  (void)ver;
+}
+
+void getSystemGuid(std::string &guid) {
+  guid = "";
+}
+
+void platformInit(uint8_t fru) {
+  (void)fru;
+}
diff --git a/crashdump.cpp b/crashdump.cpp
index 4acf161..1fb91c5 100644
--- a/crashdump.cpp
+++ b/crashdump.cpp
@@ -32,10 +32,15 @@
 #include <sys/mman.h>
 #include <sys/stat.h>
 #include <sys/types.h>
+#ifndef NO_SYSTEMD
 #include <systemd/sd-id128.h>
+#endif
 #include <termios.h>
 #include <time.h>
 #include <unistd.h>
+#ifdef NO_SYSTEMD
+#include <CLI/CLI.hpp>
+#endif
 
 #include <chrono>
 #include <fstream>
@@ -63,6 +68,10 @@ extern "C" {
 #include "crashdump.hpp"
 #include "utils_triage.hpp"
 
+#ifdef NO_SYSTEMD
+extern void platformInit(uint8_t);
+#endif
+
 #ifdef PMT_NAC_CRASHLOG
 #include "pmt_crashlog.hpp"
 #endif
@@ -70,7 +79,9 @@ extern "C" {
 namespace crashdump
 {
 static CPUInfo cpus[MAX_CPUS];
+#ifndef NO_SYSTEMD
 static std::shared_ptr<sdbusplus::asio::dbus_interface> logIface;
+#endif
 constexpr char triggerTypeOnDemand[MAX_TRIGGER_STR_LEN] = "On-Demand";
 int sectionFailCount = 0;
 static constexpr const int peciCheckInterval = 10;
@@ -84,6 +95,7 @@ int scandir_filter(const struct dirent* dirEntry)
                     crashdumpPrefix.size()) == 0);
 }
 
+#ifndef NO_SYSTEMD
 void dbusRemoveOnDemandLog()
 {
     // Always make sure the D-Bus properties are removed
@@ -196,6 +208,7 @@ void dbusAddLog(char** logContents, const std::string& timestamp,
     logIface->register_property("Filename", filename);
     logIface->initialize();
 }
+#endif
 
 void executeBAFI(char** contents, cJSON** root)
 {
@@ -210,6 +223,7 @@ void executeTriage(char** contents, cJSON** root)
 #endif
 }
 
+#ifndef NO_SYSTEMD
 void newOnDemandLog(CPUInfo* cpus, char** onDemandLogContents,
                     std::string& timestamp)
 {
@@ -408,6 +422,7 @@ void incrementCrashdumpCount()
         "org.freedesktop.DBus.Properties", "Get",
         "xyz.openbmc_project.Control.Processor.ErrConfig", "CrashdumpCount");
 }
+#endif
 
 void dbusAddStoredLog(char** storedLogContents, const std::string& timestamp,
                       char const* crashdumpStatusStr, char const* modifier)
@@ -493,6 +508,7 @@ void dbusAddStoredLog(char** storedLogContents, const std::string& timestamp,
             CRASHDUMP_PRINT(ERR, stderr, "failed to remove %s: %s\n",
                             filename.c_str(), ec.message().c_str());
         }
+#ifndef NO_SYSTEMD
         // Now remove the interface for the deleted log
         auto eraseit =
             std::find_if(storedLogIfaces.begin(), storedLogIfaces.end(),
@@ -511,6 +527,7 @@ void dbusAddStoredLog(char** storedLogContents, const std::string& timestamp,
             crashdump::server->remove_interface(iface);
             storedLogIfaces.erase(eraseit);
         }
+#endif
     }
 
     // Create the new crashdump filename
@@ -564,6 +581,7 @@ void dbusAddStoredLog(char** storedLogContents, const std::string& timestamp,
 
     // Make dbusAddStoredLog() partially unit-testable
 #ifdef BUILD_MAIN
+#ifndef NO_SYSTEMD
     // Add the new interface for this log
     std::filesystem::path path =
         std::filesystem::path(crashdumpPath) / std::to_string(crashdumpNum);
@@ -586,8 +604,10 @@ void dbusAddStoredLog(char** storedLogContents, const std::string& timestamp,
     // Increment the count for this completed crashdump
     incrementCrashdumpCount();
 #endif
+#endif
 }
 
+#ifndef NO_SYSTEMD
 void sendSdBusSignal(char const* signalName,
                      std::shared_ptr<sdbusplus::asio::dbus_interface> ifaceLog)
 {
@@ -602,6 +622,7 @@ void sendSdBusSignal(char const* signalName,
                         signalName);
     }
 }
+#endif
 
 void newStoredLog(CPUInfo* cpus, char** storedLogContents,
                   const std::string& triggerType, std::string& timestamp)
@@ -691,6 +712,7 @@ bool isPECIAvailable()
     return true;
 }
 
+#ifndef NO_SYSTEMD
 /** Exception for when a log is attempted while power is off. */
 struct PowerOffException final : public sdbusplus::exception_t
 {
@@ -733,8 +755,10 @@ struct LogInProgressException final : public sdbusplus::exception_t
         return EBUSY;
     }
 };
+#endif
 } // namespace crashdump
 
+#ifndef NO_SYSTEMD
 static void signalHandler(int sig)
 {
     CRASHDUMP_PRINT(INFO, stderr,
@@ -819,6 +843,7 @@ static void peciAvailableCheck()
         });
     }
 }
+#endif
 
 int exec_main(int enable_i3c)
 {
diff --git a/crashdump.hpp b/crashdump.hpp
index da77e67..0f0cf59 100644
--- a/crashdump.hpp
+++ b/crashdump.hpp
@@ -26,7 +26,9 @@
 #include <boost/asio/io_service.hpp>
 #include <boost/asio/steady_timer.hpp>
 #include <filesystem>
+#ifndef NO_SYSTEMD
 #include <sdbusplus/asio/object_server.hpp>
+#endif
 #include <vector>
 
 extern "C" {
@@ -58,12 +60,14 @@ constexpr char const* crashdumpTelemetryInterface =
     "com.intel.crashdump.Telemetry";
 constexpr int numStoredLogs = 3;
 
+#ifndef NO_SYSTEMD
 static boost::asio::io_service io;
 static std::shared_ptr<sdbusplus::asio::connection> conn;
 static std::shared_ptr<sdbusplus::asio::object_server> server;
 static std::vector<
     std::pair<std::string, std::shared_ptr<sdbusplus::asio::dbus_interface>>>
     storedLogIfaces;
+#endif
 
 const std::filesystem::path crashdumpDir = "/tmp/crashdump/output";
 const std::string crashdumpFileRoot{"crashdump_ondemand_"};
@@ -85,8 +89,10 @@ void dbusAddStoredLog(const std::string& storedLogContents,
 void newStoredLog(CPUInfo* cpus, std::string& storedLogContents,
                   const std::string& triggerType, std::string& timestamp);
 bool isPECIAvailable();
+#ifndef NO_SYSTEMD
 void sendSdBusSignal(char const* signalName,
                      std::shared_ptr<sdbusplus::asio::dbus_interface> ifaceLog);
+#endif
 #ifndef BUILD_MAIN
 extern int sectionFailCount;
 #endif
diff --git a/utils_common_dbusplus.cpp b/utils_common_dbusplus.cpp
index 54bcd55..e8df6d1 100644
--- a/utils_common_dbusplus.cpp
+++ b/utils_common_dbusplus.cpp
@@ -22,13 +22,23 @@ extern "C" {
 #include "engine/utils.h"
 }
 
+#ifdef NO_SYSTEMD
+extern void getSystemGuid(std::string &);
+#endif
+
 using PropertyValue =
     std::variant<bool, uint8_t, int16_t, uint16_t, int32_t, uint32_t, int64_t,
                  uint64_t, double, std::string>;
 
+#ifdef NO_SYSTEMD
+extern void getBmcVersion(char *, size_t);
+extern void getBiosVersion(char *, size_t);
+#endif
+
 namespace crashdump
 {
 
+#ifndef NO_SYSTEMD
 int getBMCVersionDBus(char* bmcVerStr, size_t bmcVerStrSize)
 {
     using ManagedObjectType = boost::container::flat_map<
@@ -178,6 +188,7 @@ std::shared_ptr<sdbusplus::bus::match::match>
             }
         });
 }
+#endif
 
 } // namespace crashdump
 
@@ -190,10 +201,14 @@ std::shared_ptr<sdbusplus::bus::match::match>
  ******************************************************************************/
 void fillBmcVersion(char* bmcVersion, size_t bmcVersionSize)
 {
+#ifdef NO_SYSTEMD
+    getBmcVersion(bmcVersion, bmcVersionSize);
+#else
     if (crashdump::getBMCVersionDBus(bmcVersion, bmcVersionSize) == ACD_FAILURE)
     {
         strcpy_s(bmcVersion, bmcVersionSize, MD_NA);
     }
+#endif
 }
 
 /******************************************************************************
@@ -205,16 +220,23 @@ void fillBmcVersion(char* bmcVersion, size_t bmcVersionSize)
  ******************************************************************************/
 void fillBiosId(char* biosVersion, size_t biosVersionSize)
 {
+#ifdef NO_SYSTEMD
+    getBiosVersion(biosVersion, biosVersionSize);
+#else
     if (crashdump::getBIOSVersionDBus(biosVersion, biosVersionSize) ==
         ACD_FAILURE)
     {
         strcpy_s(biosVersion, biosVersionSize, MD_NA);
     }
+#endif
 }
 
 void getUuid(char* platformName)
 {
     std::string ret;
+#ifdef NO_SYSTEMD
+    getSystemGuid(ret);
+#else
     sd_id128_t appId = SD_ID128_MAKE(e0, e1, 73, 76, 64, 61, 47, da, a5, 0c, d0,
                                      cc, 64, 12, 45, 78);
     sd_id128_t machineId = SD_ID128_NULL;
@@ -228,5 +250,6 @@ void getUuid(char* platformName)
         ret.insert(18, 1, '-');
         ret.insert(23, 1, '-');
     }
+#endif
     strcpy_s(platformName, MAX_PLATFORM_NAME_LEN, ret.c_str());
 }
diff --git a/utils_common_dbusplus.hpp b/utils_common_dbusplus.hpp
index 9219027..4a4f7fe 100644
--- a/utils_common_dbusplus.hpp
+++ b/utils_common_dbusplus.hpp
@@ -27,14 +27,18 @@ extern "C" {
 #include "crashdump.hpp"
 
 #include <boost/container/flat_map.hpp>
+#ifndef NO_SYSTEMD
 #include <sdbusplus/asio/object_server.hpp>
 #include <sdbusplus/bus.hpp>
 #include <sdbusplus/message.hpp>
+#endif
 #include <variant>
 
 namespace crashdump
 {
+#ifndef NO_SYSTEMD
 std::shared_ptr<sdbusplus::bus::match::match>
     startHostStateMonitor(std::shared_ptr<sdbusplus::asio::connection> conn,
                           PlatformState* platformState);
+#endif
 } // namespace crashdump
-- 
2.25.1

